File: c/mt76_csi_session.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
#include "mt76_csi_session.h"
#include <string.h>
#include <time.h>

static uint64_t now_ns(void) {
    struct timespec now; clock_gettime(CLOCK_MONOTONIC,&now);
    return (uint64_t)now.tv_sec*UINT64_C(1000000000)+(uint64_t)now.tv_nsec;
}
static int context(mt76_session_t *session, mt_session_stats_t *st) {
    if (!session) return -1;
    mt_session_snapshot(session,st);
    return st->state != MT_SESSION_RUNNING || !st->channel_known || st->band != 1 ||
           st->control != 36 || st->center != 36 || st->width_mhz != 20 ? -1 : 0;
}
static int send(mt7921_dev_t *dev, int action, const mt_csi_capture_t *capture) {
    uint8_t request[16], reply[256]; uint32_t len=sizeof(reply),status;
    unsigned receivers=action==MT_CSI_RECEIVER_COUNT ? capture->receivers : 0;
    const uint8_t *ta=action==MT_CSI_ADD_TRANSMITTER ? capture->transmitter : NULL;
    int size=mt_csi_request(dev->usb.chip,action,receivers,ta,request,sizeof(request));
    if (size<0 || mt7921_uni_option(dev->mcu.prof,0x4a,false)!=7) return -1;
    if (mt7921_mcu_uni(&dev->mcu,0x4a,request,(uint32_t)size,true,reply,&len,1000) ||
        mt_csi_ack(dev->usb.chip,reply,len,dev->mcu.msg_seq,&status) || status) return -1;
    return 0;
}
static int begin(mt7921_dev_t *dev, void *ctx) {
    mt_csi_capture_t *capture=ctx;
    mt_session_stats_t current;
    if (dev->usb.chip != MT_CHIP_MT7925 || context(capture->session,&current) ||
        current.epoch_ns!=capture->epoch_ns || current.generation!=capture->generation) return -1;
    capture->active=capture->needs_reload=true; capture->ready=false;
    const int actions[]={MT_CSI_STOP,MT_CSI_BEACON_SELECTOR,MT_CSI_START,
                         MT_CSI_ADD_TRANSMITTER,MT_CSI_RECEIVER_COUNT};
    for (unsigned i=0;i<sizeof(actions)/sizeof(*actions);i++)
        if (send(dev,actions[i],capture)) return -1;
    capture->configured_ns=now_ns();
    return 0;
}
int mt_csi_capture_start(mt76_session_t *session, mt_csi_capture_t *capture,
                          const uint8_t transmitter[6], unsigned receivers) {
    if (!capture || capture->active || mt_session_chip(session)!=MT_CHIP_MT7925) return -1;
    uint8_t request[16];
    if (mt_csi_request(MT_CHIP_MT7925,MT_CSI_ADD_TRANSMITTER,0,transmitter,request,sizeof(request))<0 ||
        mt_csi_request(MT_CHIP_MT7925,MT_CSI_RECEIVER_COUNT,receivers,NULL,request,sizeof(request))<0) return -1;
    mt_session_stats_t st;
    if (context(session,&st)) return -1;
    capture->session=session; capture->epoch_ns=st.epoch_ns; capture->generation=st.generation;
    capture->receivers=(uint8_t)receivers; memcpy(capture->transmitter,transmitter,6);
    capture->ready=false;
    int result=mt_session_call(session,begin,capture,6000,false);
    if (!result) capture->ready=true;
    return result;
}
static int stop(mt7921_dev_t *dev, void *ctx) { return send(dev,MT_CSI_STOP,ctx); }
int mt_csi_capture_stop(mt_csi_capture_t *capture) {
    if (!capture) return -1;
    capture->ready=false;
    if (!capture->active) return 0;
    int result=mt_session_call(capture->session,stop,capture,2000,false);
    if (!result) capture->active=false;
    return result;
}
int mt_csi_capture_accept(mt_csi_capture_t *capture, const mt_session_packet_t *packet,
                           mt_beacon_csi_report_t *out) {
    if (!capture || !packet || !out || packet->len>MT_SESSION_PACKET_CAPACITY) return -1;
    if (!capture->ready) return 1;
    mt_session_stats_t current;
    if (context(capture->session,&current) || current.epoch_ns!=capture->epoch_ns ||
        current.generation!=capture->generation) { capture->ready=false; return -1; }
    if (packet->epoch_ns!=capture->epoch_ns || packet->generation!=capture->generation ||
        packet->transitioning || packet->received_ns<capture->configured_ns) return 1;
    if (packet->kind!=MT_PACKET_REPLY || packet->len<44 || packet->raw[36]!=0x4a) return 1;
    mt_beacon_csi_report_t report;
    if (mt_beacon_csi_parse(MT_CHIP_MT7925,packet->raw,packet->len,&report)) return -1;
    if (memcmp(report.transmitter,capture->transmitter,6) || report.rx_index>=capture->receivers) return 1;
    *out=report;
    return 0;
}

File: mt76_csi_session.py
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Experimental, session-bound lifetime for the pinned beacon CSI profile.

The acquisition session owns USB and bounded queues; this object owns only CSI
configuration and host acceptance. The caller routes packets and serializes all
capture methods. Use exactly one CSI controller per device, no out-of-band CSI
commands or retunes, and stop CSI before stopping/destroying its session.
"""

from __future__ import annotations

import time

from mt76_csi import CsiAction, build_csi_request, parse_beacon_csi, parse_csi_ack


class CsiCaptureError(RuntimeError):
    """CSI lifetime invalid; stop the session and freshly bring up the device."""


class BeaconCsiCapture:
    """One selected transmitter on MT7925 channel36/20MHz, receivers1 or2.

    Construct/start/accept/stop on one control thread. Session must outlive this
    capture. start and stop submit bounded callbacks; do not call them from a
    session callback. A failed start may have changed firmware: active remains
    true, ready false, and needs_reload true. stop disables acceptance immediately,
    even if its firmware command fails. It does not clear needs_reload: explicitly
    reload firmware after the experiment, after the USB worker has stopped.
    """

    def __init__(self, session, transmitter: bytes, *, receivers: int = 1):
        chip = session.dev.CHIP
        # Validate the entire configuration before any command can be queued.
        self._requests = (
            build_csi_request(chip, CsiAction.STOP),
            build_csi_request(chip, CsiAction.BEACON_SELECTOR),
            build_csi_request(chip, CsiAction.START),
            build_csi_request(chip, CsiAction.ADD_TRANSMITTER, transmitter=transmitter),
            build_csi_request(chip, CsiAction.RECEIVER_COUNT, receivers=receivers),
        )
        self.session = session
        self._transmitter = transmitter
        self.receivers = receivers
        self.active = self.ready = self.needs_reload = False
        self.epoch_ns = self.channel_generation = self.configured_ns = 0

    def _context(self):
        snapshot = self.session.snapshot()
        if snapshot["state"] != "running" or snapshot["requested_channel"] != ("5GHz", 36, 36, 20):
            raise CsiCaptureError("CSI requires a running channel36/20MHz session")
        return snapshot

    @staticmethod
    def _send(dev, payload):
        if dev.uni_option(0x4A) != 7:
            raise CsiCaptureError("CSI requires SET_ACK option7")
        raw = dev.mcu_uni(0x4A, payload, timeout=1000)
        if parse_csi_ack(dev.CHIP, raw, dev.msg_seq):
            raise CsiCaptureError("CSI control rejected")

    def start(self):
        if self.active:
            raise CsiCaptureError("CSI capture already active")
        before = self._context()

        def begin(dev):
            current = self._context()
            if any(current[key] != before[key] for key in ("epoch_ns", "channel_generation")):
                raise CsiCaptureError("CSI context changed before configuration")
            self.active = self.needs_reload = True
            self.ready = False
            self.epoch_ns = current["epoch_ns"]
            self.channel_generation = current["channel_generation"]
            for payload in self._requests:
                self._send(dev, payload)
            self.configured_ns = time.monotonic_ns()

        self.session.call(begin, timeout=6)
        # Publish readiness only after the worker has completed its own deadline check.
        self.ready = True

    def stop(self):
        self.ready = False
        if not self.active:
            return
        self.session.call(lambda dev: self._send(dev, self._requests[0]), timeout=2)
        self.active = False

    def accept(self, packet):
        """Return an owned report or None for filtered/non-CSI/late packets.

        Never reads USB or removes another consumer's queue entry. Parse errors
        raise ValueError; session/epoch/retune invalidation raises CsiCaptureError
        and disables further acceptance. Preserve loss counters from the session.
        """
        if not self.ready:
            return None
        try:
            current = self._context()
            if (
                current["epoch_ns"] != self.epoch_ns
                or current["channel_generation"] != self.channel_generation
            ):
                raise CsiCaptureError("CSI epoch or channel generation changed")
        except CsiCaptureError:
            self.ready = False
            raise
        if (
            packet.epoch_ns != self.epoch_ns
            or packet.channel_generation != self.channel_generation
            or packet.transitioning
            or packet.received_ns < self.configured_ns
        ):
            return None
        raw = packet.raw
        if packet.kind != "reply" or len(raw) < 44 or raw[36] != 0x4A:
            return None
        report = parse_beacon_csi("mt7925", raw)
        if report.transmitter != self._transmitter or report.rx_index >= self.receivers:
            return None
        return report
